File: skills/characterization-methodology/references/golden-in-process.py
# Characterization pattern: in-process call
# Covers: TensorFlow → JAX, or any Python → Python refactor/migration
#
# The test harness and the legacy code are both Python, so the legacy function
# is imported directly from its source file. No wrapper needed.
# Use allclose for numerical output rather than exact equality.

import json
import sys
import numpy as np
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parents[3] / "legacy" / "src"))
from compute import compute  # the actual legacy Python function

GOLDEN = Path(__file__).parent / "golden" / "compute_output.json"


def test_characterizes_compute_output():
    result = compute(x=1.0, n_steps=10)
    actual = {"output": float(result)}  # convert to JSON-serialisable type

    if not GOLDEN.exists():  # first run: save the reference
        GOLDEN.parent.mkdir(parents=True, exist_ok=True)
        GOLDEN.write_text(json.dumps(actual, indent=2))
        return

    expected = json.loads(GOLDEN.read_text())
    np.testing.assert_allclose(actual["output"], expected["output"], rtol=1e-6)

File: skills/characterization-methodology/references/golden-in-process.jl
# Characterization pattern: in-process call via PyCall
# Covers: Python → Julia migrations
#
# The test harness lives in Julia (the target language), so characterization
# and new implementation tests share the same infrastructure.
# PyCall calls the actual Python legacy source directly — no wrapper needed.
# The same golden files are later used to verify the new Julia implementation.

using Test, JSON3, PyCall

# Point Python's module search path at the legacy source tree
pushfirst!(PyVector(pyimport("sys")."path"), joinpath(@__DIR__, "../../../legacy/src"))

const legacy = pyimport("compute")  # the actual legacy Python module

const GOLDEN = joinpath(@__DIR__, "golden", "compute_output.json")

@testset "characterize Python compute output" begin
    result = legacy.compute(x=1.0, n_steps=10)
    actual = Dict("output" => Float64(result))

    if !isfile(GOLDEN)  # first run: save the reference
        mkpath(dirname(GOLDEN))
        open(GOLDEN, "w") do f
            JSON3.write(f, actual)
        end
        return
    end

    expected = JSON3.read(read(GOLDEN, String))["output"]
    @test isapprox(actual["output"], expected; rtol=1e-6)
end
